Day03/phase1 (#50)
42-AI/bootcamp_data-engineering: day03/ex00/ex00.md

# Exercise 00 - AWS setup.

|                         |                    |
| -----------------------:| ------------------ |

> |   Turnin directory :    |  ex00              |

Reviewer: ```suggestion
|   Turn-in directory:    |  ex00              |
```

> |   Files to turn in :    |                    |

Reviewer: ```suggestion
|   Files to turn in:     |                    |
```

> |   Forbidden function :  |  None              |

Reviewer: ```suggestion
|   Forbidden function:   |  None              |
```

> |   Remarks :             |  n/a               |

Reviewer: ```suggestion
|   Remarks:              |  n/a               |
```

> ## What is the Cloud ?

Reviewer: ```suggestion
## What is the Cloud?
```

> Cloud computing is the on-demand delivery of IT resources and applications via the Internet with pay-as-you-go pricing. In reality, a cloud server is in a data center all around the world.

Reviewer: Did I get the meaning right?
```suggestion
Cloud computing is the on-demand delivery of IT resources and applications via the Internet with pay-as-you-go pricing. In fact, a cloud server is located in a data center that could be anywhere in the world.
```

Whether you run applications that share photos to millions of mobile users or deliver services that support the critical operations of your business, the cloud provides rapid access to flexible and low-cost IT resources. With cloud computing, you don’t need to make large up-front investments in hardware and spend a lot of time managing that hardware. Instead, you can provision exactly the right type and size of computing resources you need to power your newest bright idea or operate your IT department. With cloud computing, you can access as many resources as you need, almost instantly, and only pay for what you use.

In its simplest form, cloud computing provides an easy way to access servers, storage, databases, and a broad set of application services over the Internet. Cloud computing providers such as AWS own and maintain the network-connected hardware required for these application services, while you provision and use what you need for your workloads.

> As seen previously, Cloud computing has some real benefits :

Reviewer: ```suggestion
As seen previously, Cloud computing provides some real benefits :
```

> - **variable expense** : we don't need to invest in huge data centers you may not use. You pay for how much you consume !

Reviewer: ```suggestion
- **Variable expense**: You don't need to invest in huge data centers you may not use at full capacity. You pay for how much you consume!
```

> - **available in minutes** : new IT resources can be accessed within minutes.

Reviewer: ```suggestion
- **Available in minutes**: New IT resources can be accessed within minutes.
```

> - **economies of scale** : because of the number of users. Cloud providers can achieve higher economies of scale translating in lower prices.

Reviewer: ```suggestion
- **Economies of scale**: A large number of users enables Cloud providers to achieve higher economies of scale translating at lower prices.
```

> - **global in minutes** : cloud architectures can be deployed really easily all around the world.

Reviewer: ```suggestion
- **Global in minutes**: Cloud architectures can be deployed really easily all around the world.
```

> Deployments using the cloud can be `all-in-cloud-based` (the entire infrastructure is in the cloud) or `hybrid` (using on premise and cloud).

Reviewer: ```suggestion
Deployments using the cloud can be `all-in-cloud-based` (the entire infrastructure is in the cloud) or `hybrid` (using on-premise and cloud).
```

## AWS global infrastructure

> Amazon Web Service (AWS) is a cloud service provider, also known as infrastructure as a service (`IaaS`). AWS is the clear market leader in this domain and offers many services compared to other cloud providers.

Reviewer: ```suggestion
Amazon Web Services (AWS) is a cloud service provider, also known as infrastructure-as-a-service (`IaaS`). AWS is the clear market leader in this domain and offers much more services compared to its competitors.
```

> AWS has some interesting proporties as :

Reviewer: ```suggestion
AWS has some interesting properties such as:
```

> - **High availability** : Any file can be accessed from anywhere

Reviewer: ```suggestion
- **High availability**: Any file can be accessed from anywhere.
```

> - **Fault tolerance** : If an AWS server fails. You can still retrieve the file (the fault tolerancy is due to redundancy)

Reviewer: ```suggestion
- **Fault tolerance**: In case an AWS server fails, you can still retrieve the files (the fault tolerance is due to redundancy).
```

> - **Scalibility** : Possibility to add more servers when needed

Reviewer: ```suggestion
- **Scalability**: Possibility to add more servers when needed.
```

> - **Elasticity** : Possibility to grow or shrink infrastructure

Reviewer: ```suggestion
- **Elasticity**: Possibility to grow or shrink infrastructure.
```

AWS provides a highly available technology infrastructure platform with multiple locations worldwide. These locations are composed of `regions` and `availability zones`.

> Each region is a separate geographic area. Each region has multiple, isolated locations known as availability zones. An availibity zone is a physical data center geographically seperated from other availibility zones (redundant power, networking and connectivity).

Reviewer: ```suggestion
Each region represents a unique geographic area. Each region contains multiple, isolated locations known as availability zones. An availability zone is a physical data center geographically separated from other availability zones (redundant power, networking, and connectivity).
```

> We can achieve high availability by deploying your application across multiple availability zones.

Reviewer: ```suggestion
You can achieve high availability by deploying your application across multiple availability zones.
```

![AWS regions](../assets/aws_regions.png){width=400px}

The `edge locations` you see on the picture are endpoints for AWS which are used for caching content (performance optimization mechanism in which data is delivered from the closest servers for optimal application performance). Typically consists of CloudFront (Amazon's content delivery network (CDN)).

## AWS account registration

> When we create a new account on AWS, we have an AWS Free Tier. It allows us to use some of AWS resources for free each month during one year.

Reviewer: ```suggestion
When you create a new account on AWS, you get an AWS Free Tier. The latter allows you to use some of the AWS resources for free each month for one year.
```

> In order to create an AWS account, we must go to the following [address](https://aws.amazon.com).

Reviewer: ```suggestion
In order to create an AWS account, you must visit the following [address](https://aws.amazon.com).
```

> Then we click on the create an account button and fill the following form.

Reviewer: ```suggestion
Then, click on the "Create an account" button and fill the following form.
```

![Account creation 1](../assets/account_creation_1.png){width=240px}

\newpage

![Account creation 2](../assets/account_creation_2.png){width=245px}

> Don't worry! Even if we enter our card number, this day is not going to cost us anything. First,  Indeed, AWS as a free tier usage which allows us to use a small amount of AWS resources for free. This will be sufficient for what we want to do during this day. At the end of your day you will need to destroy your infrastructure entirely (don't keep things runing) !!!

Reviewer: ```suggestion
Don't worry! Even if you enter your card number, this day is not going to cost you anything. First, indeed, AWS as a free tier usage that allows you to use a small amount of AWS resources for free. This will be sufficient enough for what you are going to do today. By the end of the day, you will have to entirely destroy your infrastructure (don't keep things running) !!!
```

> In the next form, we choose the `basic plan`, which is free.

Reviewer: ```suggestion
In the next form, choose the `basic plan`, which is free.
```

> To connect to aws we use our email and password. We will then access to your root account. The root account **MUST NOT** be used directly when using AWS resources for security purposes. We will later create IAM users to use AWS resources more safely.

Reviewer: ```suggestion
To connect to AWS, use your email and password. You will then access to your root account. The root account **MUST NOT** be used directly when using AWS resources for security purposes. You will later create IAM users to use AWS resources in a safer way.
```

## Discover AWS console

> The console allows to search for specific services. by default they are sorted by group but can be sorted alphabetically.

Reviewer: ```suggestion
The console allows searching for specific services. By default, services are sorted by groups but they can also be sorted alphabetically.
```

![AWS console](../assets/console.png){width=600px}

> We notice there is a link for resource group. Resource groups allow us to take a collection of aws resources and assign a tag (a label) to them so we can manage them as a group.

Reviewer: ```suggestion
You will notice there is a link for a resource group. Resource groups allow you to take a collection of AWS resources and assign them a tag (a label) so they can be managed as a group.
```

> The pushpin is for one-click navigation. It allows us to create shortcuts for most commonly used resources.

Reviewer: ```suggestion
The pushpin is for one-click navigation. It allows you to create shortcuts for the most commonly used resources.
```

![AWS pushpin](../assets/pushpin.png){width=600px}

> The alarm icon allows us to see all system alerts. If we click we will see more details about the issues and their current status.

Reviewer: ```suggestion
The alarm icon enables you to see all system alerts. Click it to see more details about the issues and their current status.
```

![AWS alerts](../assets/alerts.png){width=200px}

> Then we have account information and the selected AWS regions we are working on (all resources are not available for every regions).

Reviewer: ```suggestion
Then, you have account information and the selected AWS regions you are working on (note that all resources are not available for every region).
```

> Support center allows us to create cases when we encounter problems (we can see support plan here, which will influence aws response time).

Reviewer: ```suggestion
The support center allows you to create cases whenever you encounter problems (you can see the support plan here, which will influence AWS response time).
```

## Setting up a billing alarm

> In order to avoid any big mistake (forget to destroy your infrastructure and it cost you some money), we are going to set an alarm to tell us if we have a cost superior to 1$.

Reviewer: ```suggestion
In order to avoid any big mistake (like forgetting to destroy your infrastructure and facing unexpected expenses), you are going to set an alarm for cost overrun. You will receive alerts when charges have exceeded 1$.
```

> We first type `billing` into aws search services bar.

Reviewer: ```suggestion
First, type `billing` into AWS search services bar.
```

![billing section](../assets/billing.png){width=450px}

> Then we can go in `Billing preferences` section and validate `Receive Free Tier Usage Alerts` (we have to enter an email). We can also set `Receive Billing Alerts` then we have to save preferences. We can now receive billing alerts !

Reviewer: ```suggestion
Then, you can go to the `Billing preferences` section and validate `Receive Free Tier Usage Alerts` (enter an email). You can also enable `Receive Billing Alerts`. Finally, save preferences. You will now receive billing alerts!
```

> Now we need to go CloudWatch to configure a billing alert. Cloudwatch is the AWS monitoring system use to track performances of used AWS resources.

Reviewer: ```suggestion
You now need to go to CloudWatch to configure a billing alert. Cloudwatch is the AWS monitoring system used to track performances of used AWS resources.
```

> We go into `CloudWatch` section. We go into the `Alarm/Billing` subsection. Then we can `create alarm`.

Reviewer: ```suggestion
Go to the `CloudWatch` section and then the `Alarm/Billing` subsection. Click `Create alarm`.
```

![alarm](../assets/account_alarm.png){width=450px}

> We have to `select metric`. Two types of metric exists : `by service` or `Total estimated charges`. We are going to select `Total estimated charges` and `USD` as currency. The threshold we want is `static` and `greater or equal` to 1 $.

Reviewer: ```suggestion
You have to `select metric`. Two types of metrics exist: `by service` or `Total estimated charges`. Select `Total estimated charges` and `USD` as currency. The threshold you will need is `static` and `greater or equal` to 1 $.
```

![alarm condition](../assets/account_alarm_cond.png){width=450px}

> On the configure actions page. we are going to keep `in alarm`. For the notification we use SNS (Simple Notification Service). We are going to `create a new topic` with the name "BillingAlarm" and we need to enter to enter our email address. Then we can create the SNS topic.

Reviewer: ```suggestion
On the configure actions page, keep `in alarm`. For the notification, use SNS (Simple Notification Service). You are going to `create a new topic` with the name "BillingAlarm" and enter your email address. You can then create the SNS topic.
```

> For the SNS topic to work, we need to confirm our email address (confirmation in the AWS email).

Reviewer: ```suggestion
So the SNS topic can work, you have to confirm your email address (confirmation in the AWS email).
```

> Then we can select `select an existing SNS topic` with "BillingAlarm".

Reviewer: ```suggestion
Choose `select an existing SNS topic` with "BillingAlarm".
```

> We can then add the name and the description of the alarm. Our alarm is finally created!

Reviewer: ```suggestion
Then, add the name and the description of the alarm. Your alarm is finally created!
```

## Create an IAM user

> AWS recommands we don't use our root account for day to day administrative tasks. For that, we are going to create a new IAM user and attach administrative access policy to it. as best practice AWS recommends we give users the minimum access to resources for users to accomplish their day to day tasks.

Reviewer: ```suggestion
AWS recommends not to use root accounts for day to day administrative tasks. Thus, you are going to create a new IAM user and attach the administrative access policy to it. As a best practice, AWS recommends granting users the minimum access to resources to accomplish their day to day tasks.
```

> First, go to the `IAM` service (IAM stands for Identity and access management). To create a user we go in the dashboard into the `users` section, then `add user`.

Reviewer: ```suggestion
First, go to the `IAM` service (IAM stands for Identity and access management). To create a user, go to the dashboard into the `users` section, then `add user`.
```

![add user](../assets/add_user1.png){width=450px}

> We can enter the name of the new user. We have to select an access type. We select both `Programmatic access` and `AWS Management Console access`. Accesses will be seen later.

Reviewer: ```suggestion
Enter the name of the new user and select an access type. Select both `Programmatic access` and `AWS Management Console access`. Accesses will be seen later.
```

> We can then select an automatically generated password or custom password (here a custom password is used) we can force the change of the password when the user logs in.

Reviewer: ```suggestion
Then, select an automatically generated password or custom password (here, a custom password is used). You are able to force the user to change of their password when they log in.
```

> The second step is to set permissions for the new user. For the first user we are going to attach an `AdministratorAccess` policy (which gives the user a full access to all AWS services).

Reviewer: ```suggestion
The second step is to set permissions for the new user. For the first one, attach an `AdministratorAccess` policy (which grants the user full access to all AWS services).
```

> Then a tag can be added to the user (optionnal). The last step reviews all the options set for the user. The user is finally created !

Reviewer: ```suggestion
Besides, a tag can be added to the user (optional). The last step is to review all the options set for the user. The user is finally created!
```

\newpage

![user review](../assets/add_user2.png){width=400px}

> At the end you will see the following screen. The access key is really important as it will be used during in the next exercise. Save the csv file otherwise you will nerver ever have access to the secret key again.

Reviewer: ```suggestion
At the end, you will see the following screen. The access key is really important as it will be used in the next exercise. Save the CSV file otherwise, you will never ever have access to the secret key again.
```

![access key](../assets/access_key.png){width=400px}

> By going in IAM, Users and clicking to our newly created user, the security credential pannel shows us the sign-in link for the user. you can use it to log to the console more safely.

Reviewer: ```suggestion
By going in IAM, Users, and clicking your newly created user, the security credential panel will show you the sign-in link for the user. Use it to log to the console more safely.
```
